packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,153 @@
+  ) {
+    return value;
+  } else if (schema.tag === "array") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return value.map((y) => {
+      return ts2aqua(y, schema.type);
+    });
+  } else if (schema.tag === "unlabeledProduct") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
 
-  if (arrowType.codomain.items.length === 1) {
-    return ts2aqua(returnValue, arrowType.codomain.items[0]);
+    return value.map((y, i) => {
+      return ts2aqua(y, schema.items[i]);
+    });
+  } else if (["labeledProduct", "struct"].includes(schema.tag)) {
+    if (typeof value !== "object" || value == null || Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return Object.entries(schema.fields).reduce((agg, [key, type]) => {
+      const val = ts2aqua(value[key], type);
+      return { ...agg, [key]: val };
+    }, {});
+  } else {
+    throw new Error("Unexpected tag: " + JSON.stringify(schema));
   }
+}
+
+export const wrapFunction = (
+  value: (...args: JSONValue[]) => JSONValue,
+  schema: ArrowWithoutCallbacks,
+): ((...args: JSONValue[]) => JSONValue) => {
+  return (...args) => {
+    const schemaArgs =
+      schema.codomain.tag === "nil" ? [] : schema.codomain.items;
+
+    const tsArgs = args.map((arg, i) => {
+      return aqua2ts(arg, schemaArgs[i]);
+    });
 
-  return arrowType.codomain.items.map((type, index) => {
-    return ts2aqua(returnValue[index], type);
-  });
+    const result = value(...tsArgs);
+    return ts2aqua(result, schema.codomain);
+  };
 };
 
-/**
- * Converts response value from aqua its representation to representation in typescript
- * @param req - call service data
- * @param arrow - aqua type definition
- * @returns response value in typescript representation
- */
-export const responseServiceValue2ts = (
-  req: CallServiceData,
-  arrow: ArrowType<any>,
-) => {
-  return match(arrow.codomain)
-    .with({ tag: "nil" }, () => {
-      return null;
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      if (x.items.length === 0) {
-        return null;
-      }
-
-      if (x.items.length === 1) {
-        return aqua2ts(req.args[0], x.items[0]);
-      }
-
-      return req.args.map((y, index) => {
-        return aqua2ts(y, x.items[index]);
-      });
-    })
-    .exhaustive();
+export const wrapServiceFunction = (
+  value: (
+    ...args: [...JSONValue[], ParticleContext]
+  ) => MaybePromise<JSONValue>,
+  schema: ArrowWithCallbacks,
+): ((
+  ...args: [...JSONValue[], ParticleContext]
+) => MaybePromise<JSONValue>) => {
+  return async (...args) => {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const jsonArgs = args.slice(0, args.length - 1) as JSONValue[];
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const context = args[args.length - 1] as ParticleContext;
```

Comment:
Any way to verify if that's actually the `ParticleContext`?

Reply:
Sometimes type assertions are necessary to avoid duplicated check that would satisfy typescript.
In cli repo I have maybe a couple of type assertions but most of the time it is possbile to write code in such a way so they are not needed.
I would suggest for **each** type assertion to add a comment that would explain why are you so confident to make this type assertion with some pointing to where you e.g. previously checked it

Reply:
Discussed in a private conversation that in this case it is in fact possbile to not use casts here if better types are used internally instead of using types that are exposed to the user where each function in the service that user registers in js has a ParticleContext provided as a last parameter